test: Add unit tests for feature ViewModels

gfrantzsword opened the pull request:
Wrote new unit tests for the ViewModels in each feature module:
- `BreedListViewModel`
- `BreedDetailViewModel`
- `FavoriteListViewModel`

gfrantzsword committed:
test: Add unit tests for feature ViewModels

Wrote new unit tests for the ViewModels in each feature module:
- `BreedListViewModel`
- `BreedDetailViewModel`
- `FavoriteListViewModel`

cmartins-sword commented on .idea/deviceManager.xml:
Please add .idea/ to your git ignore to remove these changes

ericcbm replied:
Still about this, there's a few more things that are recommended to add to gitignore to avoid more files to be sent to remote.
Check this [repo](https://github.com/github/gitignore) out, it's a collection of gitignores templates mantained by github.

gfrantzsword replied:
Thanks Eric, I'll take a look and see what works best for this project 🙂

ericcbm replied:
Please note you don't need to put on gitignore only the files you currently have and don't want to send.
Keeping all recommended will help you to avoid accidentally add them to remote when eventually added to the project.

gfrantzsword replied:
Addressed 👍 Ended up using a mix of the 'Android' and 'JetBrains' templates for better consistency across multiple environments.

cmartins-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,166 @@
+    @Before
+    fun setup() {
+        Dispatchers.setMain(testDispatcher)
+        breedRepository = mockk(relaxed = true)
```
You can use the @MockK annotation

cmartins-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,166 @@
+        coEvery { breedRepository.getBreedById(breedId) } returns targetBreed
+        every { breedRepository.getFavoriteBreeds() } returns flowOf(emptyList())
+
+        val viewModel = BreedDetailViewModel(breedRepository, savedStateHandle)
```
To avoid having to initialize your test subject on every test use:

```
private val underTest: BreedDetailViewModel by lazy {
        spyk(
            objToCopy =
                BreedDetailViewModel(
                    breedRepository = breedRepository,
                    savedStateHandle = savedStateHandle,
                ),
        )
    }
```

cmartins-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,166 @@
+
+    // CASE: everything is fine and load ok
+    @Test
+    fun initWithValidIdLoadsBreed() = runTest {
```
Please name the tests using the following convention: `WHEN ..... SHOULD ....` and divide the test in given, when, then blocks

gfrantzsword replied:
Submitted updated version for all View Models tests

cmartins-sword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListViewModel.kt:
```diff
@@ -109,14 +109,6 @@ class BreedListViewModel @Inject constructor(
                     breedRepository.addBreedToFavorites(breedId)
                 }
-
-                val newFavorites = if (currentFavorites.contains(breedId)) {
```
Why did you remove this code?

gfrantzsword replied:
Thank you for noticing this. It was a mistake on my part. It was a quick attempt on fixing a known bug where changes made to the ‘favorites’ state on the details page wouldn’t be updated when going back to the regular or favorites list, and I forgot to revert the change.

cmartins-sword commented on features/favorite_list/src/main/java/com/example/catbreeds/favorite_list/FavoriteListViewModel.kt:
```diff
@@ -48,7 +48,6 @@ class FavoriteListViewModel @Inject constructor(
         viewModelScope.launch {
             try {
                 breedRepository.removeBreedFromFavorites(breedId)
-                _favoriteBreeds.value = _favoriteBreeds.value.filter { it.id != breedId }
```
Same here

gfrantzsword replied:
Same as [this](https://github.com/gfrantzsword/CatBreeds/pull/4#discussion_r2425806562) (accidental and already reverted)

gfrantzsword committed:
revert: accidental changes to favorites state and update gitignore

- Reverted accidental changes made to the 'BreedList' and 'Favorites' on features responsible for updating the 'favoriteBreeds' state
- Updated the .gitignore to cover the entire ''.idea/' folder

gfrantzsword committed:
refactor: updated view model tests to use 'WHEN SHOULD' and 'GIVEN WHEN THEN' conventions

- Updated tests naming to follow the 'WHEN SHOULD' convention
- Divided tests into GIVEN WHEN THEN blocks

ericcbm commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,169 @@
+    @get:Rule
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+    private val testDispatcher = StandardTestDispatcher()
```
All the logic to override the Main dispatcher can be centralized into a reusable JUnit4 TestRule. Then, be reused on this and others view model tests.
https://developer.android.com/kotlin/coroutines/test#setting-main-dispatcher

ericcbm commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,169 @@
+
+        // WHEN
+        val viewModel = BreedDetailViewModel(breedRepository, savedStateHandle)
+        testDispatcher.scheduler.advanceUntilIdle()
```
just calling `advanceUntilIdle()` is enough as the `runTest` is already running on the test dispatcher.
Same for all other calls.
```suggestion
        advanceUntilIdle()
```

gfrantzsword committed:
refactor(test): Simplify advanceUntilIdle calls

Remove redundant `testDispatcher.scheduler`

gfrantzsword committed:
refactor(test): Create reusable MainDispatcher on new :test_core module

- Extracted the '@Before'/'@After' implementation to unified MainDispatcherRule.kt
- Created new :test_core module to house the new unified rule and future common test implementations
- Updated ViewModel tests to use the new rule

gfrantzsword committed:
fix(test): corrected variable placement

- Corrected the placement of 'description', 'origin' and 'life_span' on mock breed data for the BreedListViewModelTest

gfrantzsword committed:
refactor(tests): Lazily initialize ViewModel in test classes

- A single 'vmUnderTest' instance is now created per test class using 'by lazy' and 'spyk()'.
- This avoids repetitive instantiation inside each test method.
- The setup for dependencies as 'SavedStateHandle' is now handled in the '@Before' block, ensuring test isolation.

gfrantzsword committed:
refactor(test): Use @MockK annotation for mock creation

- Automates mock initialization by adding the `MockKRule` and using the `@RelaxedMockK` annotation.

gfrantzsword committed:
refactor: Update .gitignore to use recommended templates

- Now uses a mix of the 'Android' and 'JetBrains' templates from 'https://github.com/github/gitignore.git'.
- Ensures better consistency across multiple development environments.

goncalo-guilherme-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+import org.junit.Test
+
+// Mock data
+object TestBreedDetailData {
```
For better organization, maybe you could move the dummy data to other file. 
And you don't need to declare a singleton for this cases. You can just declare the val's and fun's directly on the file's body, without any object or class declaration.

gfrantzsword replied:
Good call. I've refactored the mock data creation into the :test_core module. Now, all tests share generic getBreed()/getBreeds() functions with customizable parameters

goncalo-guilherme-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+        val targetBreed = TestBreedDetailData.testBreed
+        val breedId = targetBreed.id
+        savedStateHandle["breedId"] = breedId
+        coEvery { breedRepository.getBreedById(breedId) } returns targetBreed
```
I've noticed that you are repeating the lines 74 to 76 on all tests. Maybe you could create a private method for each one of them and call those methods when needed? It would make the code a little bit clean.

gfrantzsword replied:
I created a few helper methods for all ViewModel tests that should make it cleaner overall (3c91484)

AlexCarvalhoSword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+        val breedId = targetBreed.id
+        savedStateHandle["breedId"] = breedId
+        coEvery { breedRepository.getBreedById(breedId) } returns targetBreed
+        every { breedRepository.getFavoriteBreeds() } returns flowOf(emptyList())
```
you could create some functions in order to not have repeatable code. 
for example, for these 
```
private fun whenGetBreedByIdReturns(result: Breed) {
     coEvery { breedRepository.getBreedById(any()) } returns result
}

private fun whenGetFavoriteBreedsReturns() {
    every { breedRepository.getFavoriteBreeds() } returns flowOf(emptyList())
}

```

gfrantzsword replied:
You can check 3c91484 for an updated version. I created helper methods for all ViewModel tests that should make it cleaner

AlexCarvalhoSword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+        savedStateHandle["breedId"] = breedId
+        coEvery { breedRepository.getBreedById(breedId) } returns targetBreed
+        every { breedRepository.getFavoriteBreeds() } returns flowOf(emptyList())
+        coEvery { breedRepository.addBreedToFavorites(breedId) } returns Unit
```
```suggestion
        coJustRun { breedRepository.addBreedToFavorites(breedId) }
```

gfrantzsword replied:
Made some changes on commit a2840ed

AlexCarvalhoSword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+        advanceUntilIdle()
+
+        // THEN (Verify)
+        coVerify(exactly = 1) { breedRepository.addBreedToFavorites(breedId) }
```
same for the verify 

```
fun verifyAddBreedToFavoritesIsCalled(breedId: String, exactly: Int = 1) {
   coVerify(exactly = exactly) { breedRepository.addBreedToFavorites(breedId) }
} 
```

AlexCarvalhoSword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Test
+    fun `WHEN toggling favorite SHOULD update state and call repository correctly`() = runTest {
```
we usually put the "ands" in uppercase 

```suggestion
    fun `WHEN toggling favorite SHOULD update state AND call repository correctly`() = runTest {
```

gfrantzsword replied:
Thanks, I wasn't aware

AlexCarvalhoSword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+        advanceUntilIdle()
+
+        // WHEN (Spam toggle)
+        vm.toggleFavorite(breedId) // add
```
just a suggestion 

```suggestion
   with(vm) {
        repeat(3) {
             toggleFavorite(breedId) 
        }
   }
```

gfrantzsword replied:
Agreed, this is much cleaner. I've applied the same pattern in a few other instances as well

goncalo-guilherme-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Test
+    fun `WHEN toggling favorite SHOULD update state and call repository correctly`() = runTest {
```
I've noticed that you tried to do 2 tests in 1 method, to add and remove the favourite. Maybe you could separate this into 2 different tests?

gfrantzsword replied:
Done. You can see the updated version on commit 62b9a4a

AlexCarvalhoSword commented on features/favorite_list/src/test/java/com/example/catbreeds/favorites/unitTests/FavoritesViewModelTest.kt:
```diff
@@ -0,0 +1,119 @@
+        // THEN
+        assertEquals(2, vm.favoriteBreeds.value.size)
+        val firstFavorite = vm.favoriteBreeds.value.first()
+        assertEquals("Siberian", firstFavorite.name)
```
on your TestFavoriteListData you should create constants for the dummies you use in order to have them in a type safe manner. 


Like
```

const val DUMMY_SIBERIAN_ID = "id1"
const val DUMMY_SIBERIAN_NAME = "Siberian"
const val DUMMY_SIBERIAN_ORIGIN  = "Russia"
// etc etc

```
There you can also have already the dummies for the full breed. 
```

val dummySiberianBreed = Breed(
     id = DUMMY_SIBERIAN_ID,
     name = DUMMY_SIBERIAN_NAME,
     // ....
) 

val dummyPersionBreed = Breed(
     // ....
) 

val dummyBreeds = listOf(dummySiberianBreed, dummyPersionBreed)
```

This will make it easier to handle the data and easier to make changes in the future if needed

gfrantzsword replied:
After centralizing the mock data in ff9f63a, using constants was the logical next step. I've implemented that in 952e057

goncalo-guilherme-sword commented on features/breed_list/src/test/java/com/example/catbreeds/breed_list/unitTests/BreedListViewModelTest.kt:
```diff
@@ -0,0 +1,216 @@
+        val breedId = testBreeds.first().id
+        every { breedRepository.getBreeds() } returns flowOf(testBreeds)
+        every { breedRepository.getFavoriteBreeds() } returns flowOf(emptyList())
+        coEvery { breedRepository.refreshBreeds() } returns Unit
```
You can use the coJustRun on every case that you return Unit

```suggestion
        coJustRun { breedRepository.refreshBreeds() } returns Unit
```

gfrantzsword replied:
Made some changes on commit a2840ed

gfrantzsword committed:
refactor(test): Centralized mock data creation

Removes duplicated mock data from individual ViewModel tests and consolidates it into a single `TestData.kt` file within the `:test_core` module.
- Introduces helper functions (`getBreed`, `getBreeds`) to act as a single source of truth for generating test models.

gfrantzsword commented:
Quick note: I tried renaming the PR branch, but doing so deleted the original and created a new one, closing this PR.

Once @cmartins-sword gave me a heads up, I quickly restored the branch and reopened the PR, so everything should be fine 😅

gfrantzsword committed:
refactor(test): Split toggleFavorite tests into two cases

Separates previous test into two:
- One verifies the addition of a favorite;
- The other verifies the removal.

gfrantzsword committed:
refactor(test): Use constants for mock data

- Create `MockData.kt` with `const val` for all test strings;
- Rename test methods with uppercase 'AND' as suggested;
- Update 'spam toggle' on tests to use simplified repeat() method.

gfrantzsword committed:
refactor(test): Use coJustRun when possible

Use coJustRun instead of coEvery when the return is Unit

gfrantzsword committed:
refactor(test): Create helper methods across ViewModel tests

Created multiple helper functions to simplify test setup and verification, avoiding repeated logic.

gfrantzsword committed:
refactor(test): Add data layer unit tests and align with defined standards

- Add unit tests for ConnectivityChecker and BreedRepository;
- Refactor them to use the ':test_core' module and helper methods.

This effectively merges the 'data-unit-tests' branch work, ensuring the entire codebase follows the same testing patterns

goncalo-guilherme-sword commented on data/src/test/java/com/example/catbreeds/data/unitTests/RepositoryTest.kt:
```diff
@@ -0,0 +1,184 @@
+        }
+    }
+
+    private fun setupRemoteService(networkBreeds: List<Breed>) {
```
Since the param value is always the same, I suggest you could remove the param from here and assign the value directly.

gfrantzsword replied:
I added a default value to the parameter
```
private fun setupRemoteService(setupBreeds: List<Breed> = networkBreeds) {
        coEvery { remoteService.getBreeds() } returns setupBreeds
```
Allowed me to simplify the current calls while maintaining some modularity. What do you think?

goncalo-guilherme-sword replied:
Seems to be good, I like this approach too.

goncalo-guilherme-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,190 @@
+
+        // THEN
+        verifyAddBreedToFavorites(breedId, 1)
+        verifyRemoveBreedFromFavorites(breedId, 0)
```
I believe you can remove this call. Here you already know that the breed is not a favourite one so, the probability of the remove part being called should be 0.

gfrantzsword replied:
You are correct. I simplified these verifications across the classes on 7c5dacb

goncalo-guilherme-sword commented on features/breed_detail/src/test/java/com/example/catbreeds/breed_detail/unitTests/BreedDetailViewModelTest.kt:
```diff
@@ -0,0 +1,190 @@
+
+        // THEN
+        verifyRemoveBreedFromFavorites(breedId, 1)
+        verifyAddBreedToFavorites(breedId, 0)
```
Same as above. In this case the breed is a favourite one so, the add breed code will not be called.

gfrantzsword replied:
Same as above, corrected on 7c5dacb

goncalo-guilherme-sword commented on features/breed_list/src/test/java/com/example/catbreeds/breed_list/unitTests/BreedListViewModelTest.kt:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    @Test
+    fun `WHEN search query is updated SHOULD filter list correctly by name origin or temperament`() = runTest {
```
I think you should create a method for each one of the cases you are trying to test. For example:

- Method 1 (search by Name segment)
- Method 2 (search by origin)
- Method 3 (search by origin)
- (and so on)

It would be far better to understand what is the purpose of the tests.

gfrantzsword replied:
Implementing this change naturally resulted in some repeated code, but I think overall it's a win on readability.
See the changes on 4700da2 and let me know what you think

goncalo-guilherme-sword replied:
Seems to be good

goncalo-guilherme-sword commented on features/breed_list/src/test/java/com/example/catbreeds/breed_list/unitTests/BreedListViewModelTest.kt:
```diff
@@ -0,0 +1,218 @@
+
+        // THEN
+        verifyAddBreedToFavorites(breedId)
+        verifyRemoveBreedFromFavorites(breedId, 0)
```
Same as on the other test class. If you know that the breed is not favourite, you don't need to check if the remove will be called.

gfrantzsword replied:
Same, corrected on 7c5dacb

goncalo-guilherme-sword commented on features/favorite_list/src/test/java/com/example/catbreeds/favorites/unitTests/FavoritesViewModelTest.kt:
```diff
@@ -0,0 +1,116 @@
+
+    @Before
+    fun setup() {
+        every { breedRepository.getFavoriteBreeds() } returns favoriteBreedsFlow
```
Maybe we can simplify the logic of setting the return of getFavoriteBreeds()

1. Remove the `favoriteBreedsFlow`. Seems unnecessary to me
2. Move this code of the line 47 to a different method that receives a list param. Something like this:
```
private fun whenGetFavoriteBreedsReturns(breedList: List<Breed>) {
        every { breedRepository.getFavoriteBreeds() } returns flowOf(breedList)
    }
```
3. On each test, on the given block, just call the method above and pass the item list that you want to mock on 
`getFavoriteBreeds()`.

gfrantzsword replied:
Added your suggestion on 94ab926

gfrantzsword committed:
Refactor(test): Removed unnecessary verification and simplified method on repository test-class

- Removed unnecessary toggle verification across ViewModel test classes;
- Simplified setupRemoteService method on RepositoryTest.kt to use default parameter.

ericcbm commented on features/breed_list/src/test/java/com/example/catbreeds/breed_list/unitTests/BreedListViewModelTest.kt:
```diff
@@ -0,0 +1,217 @@
+        setupRepository(emptyList())
+
+        // WHEN
+        val vm = vmUnderTest
```
The new val is not needed. We just need to call `vmUnderTest` to initialize it, because it's lazy. Then we can use aways `vmUnderTest` class member variable.

```suggestion
        vmUnderTest
```

gfrantzsword replied:
Good idea. Made the changes on 7697e0f

ericcbm commented on features/breed_list/src/test/java/com/example/catbreeds/breed_list/unitTests/BreedListViewModelTest.kt:
```diff
@@ -0,0 +1,217 @@
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+    @get:Rule
+    val mainDispatcherRule = MainDispatcherRule()
+    @get:Rule
+    val mockkRule = MockKRule(this)
```
This is a cool Rule. I din't now it existed to automatically init and release the mocks. 😉

gfrantzsword committed:
Refactor(test): Divided monolithic search-query test method into multiple smaller ones

- Divided the method into separate test cases for each scenario

gfrantzsword committed:
Refactor(test): Simplify test class by removing repeated val initialization

- Removes repeated 'val vm = vmUnderTest' to use directly 'vmUnderTest' property

gfrantzsword committed:
refactor(test): Simplify FavoriteListViewModelTest setup

- Created new helper method
to set up mocks within each test;
- Removed repeated 'val vm = vmUnderTest' to use directly 'vmUnderTest' property

goncalo-guilherme-sword reviewed (approved):
:rocket:
